### ISSUE TITLE
RemoveIPC prematurely removes IPC segments

### ISSUE THREAD

**Issue by jaymzh** [systemd/systemd#4532]
### Submission type

  - [X] Bug report
  - [ ] Request for enhancement (RFE)

### systemd version the issue has been seen with

> 231

### Used distribution

> CentOS 7.2

### In case of bug report: Expected behaviour you didn't see

>

### In case of bug report: Unexpected behaviour you saw
When `RemoveIPC` is set and postgresql is running as the postgresql user (and thus has shared-memory segments created and in use) and then someone else 'su's to the postgresql user, when they log out the shared memory is removed and postgresql dies.

`RemoveIPC` should probably only remove shared memory when the last instance of that user logs out.

### In case of bug report: Steps to reproduce the problem

1. Login is a user, create IPC
1a. Create shared memory segment
2. Login a second time as the same user, log out
3. Notice shared memory is gone.

**Comment by evverx** [comment]
Looks like a duplicate of https://github.com/systemd/systemd/issues/2039

**Comment by jaymzh** [comment]
Hey @poettering - I read through the other bug... so I **want** `RemoveIPC` - but why remove IPC fo users **currently logged in**? This seems like a bug - or possibly there's some way I can tell systemd that my postgresql user is a "system user"? Currently I have to turn it off, but this seems like not the desired situation :)

**Comment by evverx** [comment]
> but why remove IPC fo users currently logged in?

Hm, I can't reproduce this

``` sh
$ grep [UG]ID_MAX config.h
#define SYSTEM_GID_MAX 999
#define SYSTEM_UID_MAX 999

$ id -u
1000

$ busctl get-property org.freedesktop.login1 /org/freedesktop/login1 org.freedesktop.login1.Manager RemoveIPC
b true

$ cat /proc/version
Linux version 3.10.0-327.36.3.el7.x86_64 (builder@kbuilder.dev.centos.org) (gcc version 4.8.5 20150623 (Red Hat 4.8.5-4) (GCC) ) #1 SMP Mon Oct 24 16:09:20 UTC 2016

$ cat test.c
#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

int main(int argc, char *argv[]) {
    shm_open("hey", O_RDWR | O_CREAT | O_EXCL, 0777);
    for (;;)
        pause();
}

$ cc -o test -lrt ./test.c

$ ls /dev/shm/hey
ls: cannot access /dev/shm/hey: No such file or directory

$ ./test &
[1] 2098

$ ls /dev/shm/hey
/dev/shm/hey
```

```
# Another terminal
ssh to the host
logout
ssh to the host
logout
ssh to the host
$ ls /dev/shm/hey
/dev/shm/hey
```

Am I missing something?

**Comment by evverx** [comment]
Oh, I called `shm_open`

> 1a. Create shared memory segment

Anyway, this works fine too

``` sh
$ cat test.c
#include <stdlib.h>
#include <sys/ipc.h>
#include <sys/shm.h>

int main(int argc, char *argv[]) {
    int shmid;
    shmid = shmget(5678, 27, IPC_CREAT | 0666);
    shmat(shmid, NULL, 0);
    for (;;) {
        pause();
    }
}

$ make test

$ cat /proc/sysvipc/shm
       key      shmid perms                  size  cpid  lpid nattch   uid   gid  cuid  cgid      atime      dtime      ctime                   rss                  swap

$ ./test &

$ cat /proc/sysvipc/shm
       key      shmid perms                  size  cpid  lpid nattch   uid   gid  cuid  cgid      atime      dtime      ctime                   rss                  swap
      5678      65536   666                    27  3947  3947      1  1000  1000  1000  1000 1478044391          0 1478044391                     0                     0
```

```
# Another terminal
ssh to the host
logout
ssh to the host
logout
ssh to the host
$ cat /proc/sysvipc/shm
       key      shmid perms                  size  cpid  lpid nattch   uid   gid  cuid  cgid      atime      dtime      ctime                   rss                  swap
      5678      32768   666                    27  3833  3833      0  1000  1000  1000  1000 1478044256 1478044258 1478044256                     0                     0
```

**Comment by poettering** [comment]
So, the RemoveIPC= logic only applies to non-system users. It is disabled entirely for system users. Your postgresql instance really should run as system user, as it is a system service, but from the above it appears it is running as normal user instead.

For regular users, logind will leave the IPC objects around exactly as long as the user is logged in at least once. And "logged in" really means "logged in": it has to have a login session registered with logind. Now, if you run some system service as a user, then that does _not_ qualify as that. It just means some process is running under the user ID, it does mean the processes were actually registered for a session with logind/pam/utmp/selinx/… or anything.

 Long story short: make sure your postgresql user is actually registered properly as system user (i.e. with a UID < 1000) instead of regular user, and the problem should go away.

Hope this makes sense, closing.
